edx_api/grades/models.py

```diff
@@ -1,44 +1,102 @@
 """
 Business objects for the Grades API
 """
-from collections import Iterable
 
-from six import python_2_unicode_compatible
+from __future__ import unicode_literals
+from six import PY2, python_2_unicode_compatible
 
+# pylint: disable=no-name-in-module, import-error
+if PY2:
+    from collections import Iterable
+else:
+    from collections.abc import Iterable
+# pylint: enable=no-name-in-module, import-error
 # pylint: disable=too-few-public-methods
 
 
-@python_2_unicode_compatible
 class CurrentGrades(object):
     """
     Current Grades object representation
     """
-    def __init__(self, current_grade_list):
+    def __init__(self, current_grade_list, specifier, keyname):
```

Comment:
`CurrentGradesByCourse` and `CurrentGradesByUser` both receive a list of `edx_api.grades.models.CurrentGrade` objects. From there, objects of those classes take these forms respectively:
```python
# CurrentGradesByCourse
self.course_id = SOME_COURSE_ID
self.current_grades = {
  SOME_USERNAME: SOME_CURRENT_GRADE, 
  # ...
}

# CurrentGradesByUser
self.username = SOME_USERNAME
self.current_grades = {
  SOME_COURSE_ID: SOME_CURRENT_GRADE, 
  # ...
}
```

Why would it be the responsibility of this common class to set the properties of these two different types of objects? Why not just have `CurrentGradesByCourse` set its own `self.course_id` and the `self.current_grade` dict according to its own specific structure, and do the same for `CurrentGradesByUser` (but setting `self.username` instead of `self.course_id`)?

Reply:
Prior to 22c4076ad5ac7ce6b5f175ffb0e6817ff97a3050, we had `CurrentGradesByCourse` and `CurrentGradesByUser` handling their own `self.current_grades` generation, but there was quite a bit of code duplication. @agrendalath also thought it might be good to try to reduce duplication by merging this into one function, so that's why we moved all of that into the common parent class. We could step back from that commit and split it back out into two functions, they will just be two very similar `__init__` functions.

I'm not dead set on either way, I agree that this is a bit more complicated, but it might improve maintainability down the line. Do you have a preference?

Reply:
I'd very strongly prefer having the logic in two separate functions. I don't expect any maintainability/extensibility benefit to outweigh the readability issues.

Reply:
@gsidebo alright, that seems fine to me, it wound up being fairly more complicated to combine the two.
I split them back into separate functions. How does it look now?